File: client/analyze-dead-code.js
#!/usr/bin/env node

const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

console.log('='.repeat(80));
console.log('🔍 React Frontend Dead Code Analysis');
console.log('='.repeat(80));

// Create reports directory if it doesn't exist
const reportsDir = path.join(__dirname, 'reports');
if (!fs.existsSync(reportsDir)) {
  fs.mkdirSync(reportsDir);
}

// Function to run command and capture output
function runCommand(command, description) {
  console.log(`\n📊 ${description}...`);
  console.log('-'.repeat(50));
  
  try {
    const output = execSync(command, { 
      cwd: __dirname, 
      encoding: 'utf8',
      stdio: 'pipe'
    });
    return output;
  } catch (error) {
    // Many tools exit with non-zero code when finding issues
    return error.stdout || error.stderr || 'No output';
  }
}

// 1. Check for unused files and dependencies
const unimportedOutput = runCommand('npx unimported --json', 'Checking for unused files and dependencies');
console.log(unimportedOutput);

try {
  const unimportedData = JSON.parse(unimportedOutput);
  
  // Write detailed report
  const report = {
    timestamp: new Date().toISOString(),
    summary: {
      unusedDependencies: unimportedData.unusedDependencies?.length || 0,
      unimportedFiles: unimportedData.unimportedFiles?.length || 0,
      unresolvedImports: unimportedData.unresolvedImports?.length || 0
    },
    details: unimportedData
  };
  
  fs.writeFileSync(
    path.join(reportsDir, 'unimported-analysis.json'),
    JSON.stringify(report, null, 2)
  );
  
  console.log(`✅ Found ${report.summary.unusedDependencies} unused dependencies`);
  console.log(`✅ Found ${report.summary.unimportedFiles} unimported files`);
  
} catch (error) {
  console.log('⚠️  Could not parse unimported output as JSON, saving raw output...');
  fs.writeFileSync(
    path.join(reportsDir, 'unimported-analysis.txt'),
    unimportedOutput
  );
}

// 2. Check dependencies with depcheck
const depcheckOutput = runCommand('npx depcheck --json', 'Checking dependencies with depcheck');
console.log('\n📊 Depcheck Results:');
console.log('-'.repeat(50));

try {
  const depcheckData = JSON.parse(depcheckOutput);
  
  console.log(`Unused dependencies: ${depcheckData.dependencies?.length || 0}`);
  console.log(`Unused devDependencies: ${depcheckData.devDependencies?.length || 0}`);
  console.log(`Missing dependencies: ${depcheckData.missing ? Object.keys(depcheckData.missing).length : 0}`);
  
  if (depcheckData.dependencies?.length > 0) {
    console.log('\nUnused dependencies:');
    depcheckData.dependencies.forEach(dep => console.log(`  - ${dep}`));
  }
  
  if (depcheckData.devDependencies?.length > 0) {
    console.log('\nUnused devDependencies:');
    depcheckData.devDependencies.forEach(dep => console.log(`  - ${dep}`));
  }
  
  fs.writeFileSync(
    path.join(reportsDir, 'depcheck-analysis.json'),
    JSON.stringify(depcheckData, null, 2)
  );
  
} catch (error) {
  console.log('⚠️  Could not parse depcheck output as JSON, saving raw output...');
  fs.writeFileSync(
    path.join(reportsDir, 'depcheck-analysis.txt'),
    depcheckOutput
  );
}

// 3. Check for unused exports
console.log('\n📊 Checking for unused exports...');
console.log('-'.repeat(50));

try {
  const tsUnusedExportsOutput = runCommand('npx ts-unused-exports tsconfig.json', 'Checking for unused exports');
  console.log(tsUnusedExportsOutput);
  
  fs.writeFileSync(
    path.join(reportsDir, 'unused-exports.txt'),
    tsUnusedExportsOutput
  );
} catch (error) {
  console.log('⚠️  ts-unused-exports not available or no tsconfig.json found');
}

// 4. Generate summary report
const summaryReport = {
  timestamp: new Date().toISOString(),
  project: 'React Frontend',
  analyzedDirectory: __dirname,
  analysisTypes: [
    'Unused files and dependencies (unimported)',
    'Dependency analysis (depcheck)',
    'Unused exports (ts-unused-exports)'
  ],
  reportsGenerated: fs.readdirSync(reportsDir)
};

fs.writeFileSync(
  path.join(reportsDir, 'summary.json'),
  JSON.stringify(summaryReport, null, 2)
);

console.log('\n🎉 Analysis Complete!');
console.log('='.repeat(80));
console.log(`📁 Reports saved to: ${reportsDir}`);
console.log(`📄 Summary report: ${path.join(reportsDir, 'summary.json')}`);
console.log('\nFiles analyzed:');
summaryReport.reportsGenerated.forEach(file => console.log(`  - ${file}`));

File: backend/analyze-dead-code.py
#!/usr/bin/env python3

import os
import json
import subprocess
import sys
from datetime import datetime

def print_header(title):
    print('=' * 80)
    print(f'🔍 {title}')
    print('=' * 80)

def print_section(title):
    print(f'\n📊 {title}...')
    print('-' * 50)

def run_command(command, description):
    """Run a command and capture output"""
    print_section(description)
    
    try:
        result = subprocess.run(
            command,
            shell=True,
            capture_output=True,
            text=True,
            cwd=os.path.dirname(os.path.abspath(__file__))
        )
        return result.stdout, result.stderr, result.returncode
    except Exception as e:
        return '', str(e), 1

def main():
    print_header('Python Backend Dead Code Analysis')
    
    # Create reports directory
    reports_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'reports')
    os.makedirs(reports_dir, exist_ok=True)
    
    analysis_results = {
        'timestamp': datetime.now().isoformat(),
        'project': 'Python Backend',
        'analyzedDirectory': os.path.dirname(os.path.abspath(__file__)),
        'tools': []
    }
    
    # 1. Run vulture for dead code detection
    stdout, stderr, returncode = run_command('vulture .', 'Scanning for dead code with vulture')
    
    vulture_issues = []
    if stdout:
        print(stdout)
        vulture_issues = stdout.strip().split('\n') if stdout.strip() else []
        
    with open(os.path.join(reports_dir, 'vulture-analysis.txt'), 'w') as f:
        f.write(f"Vulture Analysis Results\n")
        f.write(f"========================\n\n")
        f.write(f"Found {len(vulture_issues)} potential dead code issues:\n\n")
        f.write(stdout)
        
    analysis_results['tools'].append({
        'name': 'vulture',
        'description': 'Dead code detection',
        'issues_found': len(vulture_issues),
        'exit_code': returncode
    })
    
    print(f"✅ Found {len(vulture_issues)} potential dead code issues")
    
    # 2. Run unimport for unused imports
    stdout, stderr, returncode = run_command('unimport --check .', 'Checking for unused imports with unimport')
    
    unimport_issues = []
    if stdout:
        print(stdout)
        unimport_issues = stdout.strip().split('\n') if stdout.strip() else []
        
    with open(os.path.join(reports_dir, 'unimport-analysis.txt'), 'w') as f:
        f.write(f"Unimport Analysis Results\n")
        f.write(f"=========================\n\n")
        f.write(f"Found {len(unimport_issues)} unused imports:\n\n")
        f.write(stdout)
        
    analysis_results['tools'].append({
        'name': 'unimport',
        'description': 'Unused imports detection',
        'issues_found': len(unimport_issues),
        'exit_code': returncode
    })
    
    print(f"✅ Found {len(unimport_issues)} unused imports")
    
    # 3. Run pyflakes for unused variables and imports
    stdout, stderr, returncode = run_command('pyflakes .', 'Checking for unused variables and imports with pyflakes')
    
    pyflakes_issues = []
    if stdout:
        print(stdout)
        pyflakes_issues = stdout.strip().split('\n') if stdout.strip() else []
        
    with open(os.path.join(reports_dir, 'pyflakes-analysis.txt'), 'w') as f:
        f.write(f"Pyflakes Analysis Results\n")
        f.write(f"=========================\n\n")
        f.write(f"Found {len(pyflakes_issues)} issues:\n\n")
        f.write(stdout)
        
    analysis_results['tools'].append({
        'name': 'pyflakes',
        'description': 'Unused variables and imports detection',
        'issues_found': len(pyflakes_issues),
        'exit_code': returncode
    })
    
    print(f"✅ Found {len(pyflakes_issues)} pyflakes issues")
    
    # 4. Generate file analysis
    print_section('Analyzing Python files in directory')
    
    python_files = []
    for root, dirs, files in os.walk('.'):
        # Skip certain directories
        dirs[:] = [d for d in dirs if d not in ['__pycache__', '.git', 'venv', 'env', 'migrations']]
        
        for file in files:
            if file.endswith('.py'):
                file_path = os.path.join(root, file)
                python_files.append(file_path)
    
    analysis_results['file_analysis'] = {
        'total_python_files': len(python_files),
        'files': python_files
    }
    
    print(f"📁 Found {len(python_files)} Python files")
    
    # 5. Save summary report
    with open(os.path.join(reports_dir, 'summary.json'), 'w') as f:
        json.dump(analysis_results, f, indent=2)
    
    # 6. Generate consolidated report
    total_issues = sum(tool['issues_found'] for tool in analysis_results['tools'])
    
    print('\n🎉 Analysis Complete!')
    print('=' * 80)
    print(f"📁 Reports saved to: {reports_dir}")
    print(f"📄 Summary report: {os.path.join(reports_dir, 'summary.json')}")
    print(f"🔍 Total issues found: {total_issues}")
    print("\nReports generated:")
    for file in os.listdir(reports_dir):
        print(f"  - {file}")
    
    # Exit with non-zero code if issues found
    if total_issues > 0:
        print(f"\n⚠️  {total_issues} potential issues found - review the reports!")
        sys.exit(1)
    else:
        print("\n✅ No issues found!")
        sys.exit(0)

if __name__ == '__main__':
    main()